*: add design doc for resource control (#41509)
pingcap/tidb: docs/design/2022-11-25-global-resource-control.md

# Global Resource Control in TiDB
- Author: [nolouch](https://github.com/nolouch), [BornChanger](https://github.com/BornChanger), [golvr](https://github.com/golvr), [Connor1996](https://github.com/Connor1996), [tiancaiamao](https://github.com/tiancaiamao)
- Tracking Issue: https://github.com/pingcap/tidb/issues/38825

## Summary
A Global Resource Control mechanism that limits multiple applications' resource usage in a shared TiDB cluster, avoids interference between applications.

## Background and Motivation
A TiDB cluster for multiple applications to share the same data service, and there are several Aurora or MySQL, or other database instances as the data service backend. TiDB has the advantage of scalability but also has some challenges in the big-size cluster. How to make the shared cluster provide stable, predictable performance for different Applications? Part of what needs to be done is to control the resource usage of each application.

> Currently, TiDB lack of global admission control limits the request from the SQL layer to the storage layer, and different users or applications may influence each other in the shared environment. Use a global admission control to ensure that the cluster doesn’t become overloaded, avoid interference between applications, and enforce the throughput limits requested by customers.

Reviewer: ```suggestion
Currently, TiDB lack of global admission control limits the request from the SQL layer to the storage layer, and different users or applications may influence each other in the shared environment. Use a global admission control to ensure that the cluster doesn't become overloaded, avoid interference between applications, and enforce the throughput limits requested by customers.
```

Moreover, this resource control mechanism can be used by Resource Manage in the future of TiDB Cloud. Users can pay for the actual usage.  So, consider scenarios. It supports two models.

- **Provisioned:** On-Premise deployment, multi-applications shared on the big cluster. In this mode, you provision the number of RU rate limits or(CPU, IO Rate) and priority for the users/tenant/sessions.
- **Serverless:** In this mode, the user can pay the total cumulative RU for the tenant or user. Then the resources can not be limited until the RU is exhausted. This requirement TiDB Cloud has sufficient flexibility. such as auto-scale resources.

## Detailed Design

![resource control layout.png](imgs/resource-control-layout.png)

**Resource Group** helps users to control over its resources, to enable or restrict resource consumption. resource groups have attributes that define the group. all attributes can be set at group creation time and  modified any time thereafter. The attributes define how many resources unit in this group relative CPU, IO, Memory and so on.

> resource groups can be assigned to different resource objects. such as session/users/ tenant.

Reviewer: ```suggestion
resource groups can be assigned to different resource objects. such as session/users/tenant.
```

> There are two dimensions to control the resources, one is Global Quota dimension, which is more of a tenant-level application, users can plan the use of resources in provision, to limit the resources usage. On the other side is the scheduling(weight ) dimension, to ensure better control behavior when the resources are sufficient.

Reviewer: ```suggestion
There are two dimensions to control the resources, one is Global Quota dimension, which is more of a tenant-level application, users can plan the use of resources in provision, to limit the resources usage. On the other side is the scheduling(weight) dimension, to ensure better control behavior when the resources are sufficient.
```

> **Global Quota dimension:**  This part is the upper limiter and lower limit(Reservation)  for the resource group. This defines the provision of the resource usage of the application layer.  Here mainly use a token bucket algorithm to limit the resources at the entry of TiDB, count total resource usage including sql, storage layer.

Reviewer: ```suggestion
**Global Quota dimension:**  This part is the upper limiter and lower limit(reservation) for the resource group. This defines the provision of the resource usage of the application layer.  Here mainly use a token bucket algorithm to limit the resources at the entry of TiDB, count total resource usage including sql, storage layer.
```

> **Scheduling(weight) dimension:**  Currently, it’s mainly used in the internal mechanism, such as inner jobs DDL, analyzed to reduce the background job influence for priority purpose, let background jobs not influence front jobs. and on the other hand, it makes sure the resource group is not abused, prevent a resource group makes one or more nodes overhead.  it may have a user interface in the future, not exposed now for simplicity and clarity . Here mainly use the mclock/dmclock algorithm at the storage layer, and use different priority queues to influence different workers of the SQL layer.

Reviewer: ```suggestion
**Scheduling(weight) dimension:**  Currently, it’s mainly used internally, for background jobs, such as inner DDL jobs or analyze jobs, to reduce the influence of front jobs. And on the other hand, it makes sure the resource group is not abused, prevent a resource group makes one or more nodes overhead.  It may have a user interface in the future, not exposed now for simplicity and clarity . Here mainly use the mclock/dmclock algorithm at the storage layer, and use different priority queues to influence different workers of the SQL layer.
```

Will introduce this two dimensions control mechanism in the following.

### User SQL interface
The `Request Unit` can be assigned to a different level, like user or tenant. The syntax is like MySQL resource group, but it is used differently:

```sql
CREATE RESOURCE GROUP name resouce_unit=value [, ... ]
/*eg:*/

> CREATE RESOURCE GROUP small_application RRU=20000, WRU=10000;

Reviewer: No `RRU/WRU`, replaced by `RU` now?

Author: I keep the original copy plus the RRU as well in the RFC. It's not completely the same as the implementation. Is it ok?

CREATE RESOURCE GROUP small_application RU_PER_SEC=20000;
/* Some request unit maybe can provide directly*/
(CPU=2 core, READ_IO_RATE=200 MB/s, WRITE_IO_RATE=100 MB/s)
/* bind resource group via DCL */
ALTER USER user_name RESOURCE GROUP resouce_group_name;

> CREATE USER user_name IDENTIFIED BY ‘password’ RESOURCE GROUP resource_group_name;

Reviewer: ```suggestion
CREATE USER user_name IDENTIFIED BY 'password' RESOURCE GROUP resource_group_name;
```

    double requested_wru = 4;
    uint64 target_request_period_seconds = 5;
}
message Consumption {
    double rru = 1;
    double wru = 2;
    uint64 read_requests = 3;
    uint64 read_bytes = 4;
    uint64 write_requests = 5;
    uint64 write_bytes = 6;
    double cpu_seconds = 6;
}
```

In this protocol, there are some key fields:
- `target_request_period_seconds`: This field is the lease interval for the local token buckets, it is configurable.
- `requested_rru/ request_wru`: It's a load expect with load estimation. This is an exponentially weighted moving average of the load at the nod.

#### Bursting and Refill rate
Bucket Tokens have two parts.
- `burst`: it is an accumulated value, it absorbs temporal spikes in workloads of the tenant user. Moreover, if the system is elastic enough, such a cluster can be auto-scale in the cloud. It can be used as the budget for users to pay. If the burst is enough, customers feel free to use resources without limitation, which matches the main idea of the serverless product.
- `refill rate`: rate-limit resource consumption. grant the tokens over a specified period of time, according to a fraction of the global refill rate

> Once a request with a token bucket server is completed, the node adjusts its local token bucket according to the response. either adding burst tokens or setting up a refill rate.

Reviewer: ```suggestion
Once a request with a token bucket server is completed, the node adjusts its local token bucket according to the response. Either adding burst tokens or setting up a refill rate.
```

#### Debt
The token buckets can be "debt", as we mentioned, due to some statistics only being accounted for after the fact (eg, the read size of coprocess request, the CPUSecs usage). When a local bucket is in debt, KV operations are blocked until the debt is paid. This may lead to the latency not being stable . There is an improvement that allows the debt to be paid over a time period from the refill rate.
#### Adaptive Capacity

> When a request from the application arrives, the request deducts tokens. Eventually, the LAC will run out of tokens because of consumption or expiry. When the LAC runs out of tokens, it requests more tokens from Global Token Server immediately. The Glocal Token Server instance uses the information provided by the client to estimate the global token consumption and vends tokens available for the next time unit to the client’s share of overall tokens. Thus, it ensures that workloads can adaptively turn the  local bucket fill rage up to the maximum capacity.

Reviewer: ```suggestion
When a request from the application arrives, the request deducts tokens. Eventually, the LAC will run out of tokens because of consumption or expiry. When the LAC runs out of tokens, it requests more tokens from Global Token Server immediately. The Glocal Token Server instance uses the information provided by the client to estimate the global token consumption and vends tokens available for the next time unit to the client's share of overall tokens. Thus, it ensures that workloads can adaptively turn the local bucket fill rage up to the maximum capacity.
```

#### Balance the consumed capacity
The data hosted on a single storage node could be wholly unrelated and belong to different users or tenants. where each application could be from a different customer and have varied traffic patterns.  co-location was a straightforward problem with provisioned performance for application. Balancing the requirements of the provision for user's data may need to be considered.

### Scheduling Control
Compared to Storage, SQL layer node is easier to scale out since it's stateless, especially with k8s.
Storage(TiKV/Tiflash) cluster sharing with QoS guarantee is more challenging. In this chapter, we will discuss our design on request scheduling at Storage side first. There are 3 goals:
- Requests from higher priority resource group are scheduled more
- No starving for lower priority request
- Soft quota limiters can be issued to resource groups

#### QoS Metadata management

> Resource Manager  keeps the persisted metadata of resource groups. Resource Manager will translate that metadata information from abstract RUs into system resources like vCPU, read bandwidth, write bandwidth, etc to TiKV/Tiflash nodes.

Reviewer: ```suggestion
Resource Manager keeps the persisted metadata of resource groups. Resource Manager will translate that metadata information from abstract RUs into system resources like vCPU, read bandwidth, write bandwidth, etc to TiKV/Tiflash nodes.
```

> Metadata sync-up is bidirectional. Resource Manager notifies TiKV on resource group metadata change at DDL execution time via http API interface. And TiKV nodes inquire  Resource Manager when local metadata is missing or might be stale.  Resource group metadata is in memory only at TiKV nodes.

Reviewer: ```suggestion
Metadata sync-up is bidirectional. Resource Manager notifies TiKV on resource group metadata change at DDL execution time via http API interface. And TiKV nodes inquire Resource Manager when local metadata is missing or might be stale.  Resource group metadata is in memory only at TiKV nodes.
```

#### Soft quota limiters
We can specify the quota limits for some resource groups to prevent the overrun and for TiKV node protection, by reusing the front-end quota limiter (token bucket algorithm) framework of TiKV. With the quota limiter capping, requests from some resource group might be throttled even when the TiKV instance is not saturated.

#### New field in kv protobuf
TiDB needs to inject into gRPC requests a resource group related information, like the resource group id,  in the protobuf message.

#### Front-end requests scheduling

> Resource consumption on vCPU, read bandwidth and write bandwidth needs to be normalized into internal “Shares” represented as a number.  At different modules and different phases, the normalization formulas are different.

Reviewer: ```suggestion
Resource consumption on vCPU, read bandwidth and write bandwidth needs to be normalized into internal "Shares" represented as a number. At different modules and different phases, the normalization formulas are different.
```

> Read request and write request handling are quite different in TiKV. Read request handling in TiKV is already in fine grain,  and we can almost directly use request elapsed time as resource consumption, while write path handling is rather complex.

Reviewer: ```suggestion
Read request and write request handling are quite different in TiKV. Read request handling in TiKV is already in fine grain, and we can almost directly use request elapsed time as resource consumption, while write path handling is rather complex.
```

##### Read path
Currently, all read requests are executed by the yatp thread pool, whose scheduler is a multi-level feedback queue, and each level is a FIFO queue.
Because the current yatp's scheduler is a FIFO queue, a new task is always put at the end of the queue, thus the latency of all tasks are high when there are a lot of tasks.

**mClock based scheduling**

> `Mclock` is a priority-based scheduling algorithm. Each resource group maintains a separate task queue, the schedule always pops and executes a task for the queue whose next task has the minimum priority value. The

Reviewer: ```suggestion
`mClock` is a priority-based scheduling algorithm. Each resource group maintains a separate task queue, the schedule always pops and executes a task for the queue whose next task has the minimum priority value. The
```

$$P_i = \max(P{i-1} + w, V_t)$$
$P_{i-1}$ is the accumulated weight of executed tasks. Vt is the minimum priority of all non-empty queues, it is used when queue i is empty.
`w` is a number that represents the resource consumed by this task.

> It's easy to prove that by using the mclock algorithm, we can ensure a resource group can obtain its share of resources with reasonable latency.

Reviewer: ```suggestion
It's easy to prove that by using the mclock algorithm, we can ensure a resource group to obtain its share of resources with reasonable latency.
```

- For fetch_fsms, fsms of high priority should be fetched first.
- High priority is set to the context of raft entry as well. So followers can know the region is of high priority.

#### Front-end requests vs back-end requests
Backend requests scheduling is also taken into account. They can be assigned to some predefined resource groups and quota limiters.  And schedulers can handle them along with front-end requests at some underneath level, where IO is more sensitive.

## Other Issues
### On-demand provisioning
Because the concept of request units is new to customers, it may be challenging to forecast the provisioned throughput. Also, the application's data distribution and workload non-uniform access may lead to not all data across all storage nodes. Although from the global perspective of the entire cluster, there remain many resources, the application setting may still be inappropriate. Hence, we should let the cluster have a capacity to burst at the node level, and it may need to do something about balancing the consumed capacity mentioned above.

> ###Diagnostics

Reviewer: ```suggestion
### Diagnostics
```

Metrics on # of blocked requests due to out of token,  per user, per tenant.
HWM of RU for slow queries, for individual users and tenants
LWM, HWM, Average rate for individual resource group

> ###Reference

Reviewer: ```suggestion
### Reference
```
